client/components/popover/menu-item.jsx

```diff
@@ -73,6 +73,7 @@ export default class PopoverMenuItem extends Component {
 				onMouseOver={ this.handleMouseOver }
 				tabIndex="-1"
 				className={ classes }
+				data-e2e-item={ children }
```

Comment:
The component can be used with arbitrary markup as content:
```js
<PopoverMenuItem>Remove <b>post</b></PopoverMenuItem>
```
Then the `children` become an opaque object and the `data-e2e-item` has value `[object Object]`.

Reply:
I don't know at the moment how to solve this. The content of the item can be arbitrary markup.

And the popover component is likely to become a part of `@automattic/components` package used across all our products, so the components should better be rock solid without any fragile edge cases.

Maybe `data-e2e-item` should be a symbolic string explicitly specified by developer rather than auto-generated from the content? The e2e tests want to use the `data-e2e-item` attribute to find a certain item, do I understand that correctly?

Reply:
> The e2e tests want to use the data-e2e-item attribute to find a certain item, do I understand that correctly?

Exactly. 

What do you think about auto-generated from the content and converted to string? In that case, if something like arbitrary markup happens, the component will work, as usual, just `data-e2e-item` won't be useful. But, in that case, it probably won't be used by e2e tests. 

Other solution could be to add data-e2e for several items which we need for tests and call specific item when needed. But, it looks to me, that solution needs more maintenance on both sides?

Reply:
@jsnajdr I like this logic and it looks useful, but I didn't manage to make it work. Would you mind to take a look at [this](https://github.com/Automattic/wp-e2e-tests/blob/add/data-e2e-selectors/lib/components/notifications-component.js#L30) and try with your solution from above? Any help is more than welcome :)

Reply:
Hi @Stojdza! I tried to make my script above actually work and discovered that it needs two changes:
- `By.css` is just a selector object that needs to be passed to `driver.findElement` or `driver.findElements` in order to actually search for elements.
- `WebElement.getText()` returns a promise, so it's not so easy to filter the array of candidate elements. We need to wait for `getText` to resolve before deciding whether to include it in the output or not. Fortunately there is a helper function `promise.filter` in the Selenium WebDriver library that makes this easy.

Here is a complete Selenium script that you can run from command line in Node:
```js
const { Builder, By, promise } = require( 'selenium-webdriver' );

var driver = new Builder().forBrowser( 'chrome' ).build();

// defines a selector function to be used in `driver.findElements`
function themeByName( themeName ) {
  return async function( driver ) {
    // find all theme title elements by CSS selector
    const allThemeTitles = await driver.findElements( By.css( '.theme__info-title' ) );
    // filter the list by matching the text against the desired one
    const themeTitles = await promise.filter(
      allThemeTitles,
      async titleEl => await titleEl.getText() === themeName
    );
    // return the array of matching elements
    return themeTitles;
  };
}

async function navigateToTheme( themeName ) {
  // open WordPress themes page
  await driver.get( 'http://wordpress.com/themes' );
  // find matching theme title
  // (`findElement` returns first element of the array returned by selector)
  const myThemeTitle = await driver.findElement( themeByName( themeName ) );
  // click on the found element
  await myThemeTitle.click();
  // wait 5 seconds
  await new Promise( r => setTimeout( r, 5000 ) );
  // close browser and exit
  await driver.quit();
}

// find theme title "Radcliffe 2" and click on it
navigateToTheme( 'Radcliffe 2' );
```

Reply:
Thanks for this @jsnajdr, it was helpful! I used `promise.filter` and it works really well. [Here](https://github.com/Automattic/wp-e2e-tests/pull/1505/commits/11b31adba3ea15c580fa30be2cfc4b2c86676fbe) is the e2e commit for that change (your previous proposal updated with `promise.filter`).

Reply:
And [here](https://github.com/Automattic/wp-e2e-tests/pull/1505/commits/b004fca04ed6fc4a5c3fd71c2146fc7b6aa6a026) is the same approach for theme action items.